src/controllers/engine/controllerengine.cpp

```diff
@@ -198,24 +197,26 @@ bool ControllerEngine::isReady() {
 
 void ControllerEngine::initializeScriptEngine() {
     // Create the Script Engine
-    m_pEngine = new QScriptEngine(this);
+    m_pEngine = new QJSEngine(this);
 
     // Make this ControllerEngine instance available to scripts as 'engine'.
-    QScriptValue engineGlobalObject = m_pEngine->globalObject();
-    engineGlobalObject.setProperty("engine", m_pEngine->newQObject(this));
+    QJSValue engineGlobalObject = m_pEngine->globalObject();
+    ControllerEngineJSProxy* proxy = new ControllerEngineJSProxy(this);
+    engineGlobalObject.setProperty("engine", m_pEngine->newQObject(proxy));
 
     if (m_pController) {
         qDebug() << "Controller in script engine is:" << m_pController->getName();
 
+        ControllerJSProxy* controllerProxy = new ControllerJSProxy(m_pController);
```

Comment:
This does not work. Before, `m_pController` was exposed to the script engine which allowed the various Controller subclasses to add their own methods for the script engine. This is required for `midi.sendShortMsg`, which is currently missing from the script environment. I see lots of errors like this in the log:
```
Debug [Controller]: "Connected ([EffectRack1_EffectUnit1_Effect1], enabled) to connection {90722931-6f57-403c-9f77-fca12b930089}"
Warning [Controller]: ControllerEngine: Invocation of connection  "{90722931-6f57-403c-9f77-fca12b930089}" "connected to ([EffectRack1_EffectUnit1_Effect1], enabled) failed:" "TypeError: Property 'sendShortMsg' of object QObject(0x7fdbb402f790) is not a function"
```
I suggest adding a method to the Controller class that returns a ControllerJSProxy. Then make a MidiControllerJSProxy subclass of ControllerJSProxy and add `sendShortMsg` and `sendSysexMsg` functions to it.

Reply:
MidiControllerJSProxy can't be a subclass because it needs a MidiController, but ControllerJSProxy has a Controller member. Any ideas how to solve this?

Reply:
I don't understand what the problem is. MidiController is a subclass of Controller, so can't you pass the MidiController* to the ControllerJSProxy constructor?

Reply:
yes, but then I have a member of type Controller, so I can't call the methods of MidiController. Shall I downcast?

Reply:
Can't you still call MidiController's methods in MidiControllerJSProxy? It's just in ControllerJSProxy where you can't access MidiController's methods.

Reply:
`m_pController` is private in ControllerJSProxy, so it shouldn't be accessible to subclasses anyway. You could store the MidiController* pointer as a different private member of MidiControllerJSProxy.

Reply:
ah! I see. Thanks

Reply:
Ok, done for MidiController. Can you do a quick test that the problem is fixed? Then I'll do the same for the other Controllers.

Reply:
Nope, still the same error. :/ I think MidiControllerJSProxy is getting downcasted in ControllerEngine::initializeScriptEngine before being inserted into the script environment:
```
        ControllerJSProxy* controllerProxy = m_pController->jsProxy();
```

Reply:
Actually the problem was simpler than that. You forgot the `Q_OBJECT` macro in the ControllerJSProxy and MidiControllerJSProxy classes.